storage: fix various problems with dynamic replication factor

petermattis opened the pull request:
* roachtest: add replicate/wide roachtest
* storage: reimplement StorePool.AvailableNodeCount
* storage: fix lastReplicaAdded computation
* storage: fix filterUnremovableReplicas badness

Fixes #34122

Release note (bug fix): Avoid down-replicating widely replicated ranges
when nodes in the cluster are temporarily down.

petermattis committed:
storage: fix filterUnremovableReplicas badness

`filterUnremovableReplicas` was allowing replicas that were a necessary
part of quorum to be removed. This occurred due to
`filterBehindReplicas` taking a "brand new replica" that was being
considered up-to-date even when we didn't have evidence of it being
up-to-date. `filterBehindReplicas` needs to return an accurate picture
of the up-to-date replicas. Rather than push this work into
`filterBehindReplicas`, `filterUnremovableReplicas` has been changed to
perform the filtering of the "brand new replica" from the removable
candidates.

See #34122

Release note: None

petermattis committed:
storage: fix lastReplicaAdded computation

Previously, `Replica.mu.lastReplicaAdded` was being set to the maximum
replica ID in the descriptor whenever the descriptor changed. This was
invalid when a replica was removed from the range. For example, consider
a range with 9 replicas, IDs 1 through 9. If replica ID 5 is removed
from the range, `lastReplicaAdded` was being set to 9. Coupled with the
bug in the previous commit, this was causing replica ID 9 to appear to
be up-to-date when it wasn't. The fix here isn't strictly necessary, but
is done to bring sanity: `lastReplicaAdded` should accurately reflect
the last replica which was added, not the maximum replica ID in the
range.

See #34122

Release note: None

petermattis committed:
storage: reimplement StorePool.AvailableNodeCount

Reimplement `StorePool.AvailableNodeCount` in terms of
`NodeLiveness.GetNodeCount`. The latter returns a count of the user's
intended number of nodes in the cluster. This is added nodes minus
decommissioning (or decommissioned) nodes.

This fixes misbehavior of the dynamic replication factor heuristic. The
previous `StorePool.AvailableNodeCount` implementation would fluctuate
depending on the number of node descriptors that had been received from
gossip and the number of dead nodes. So if you had a 5-node cluster and
2 nodes died for more than 5 min (and thus marked as dead), the cluster
would suddenly start down-replicating ranges. Similarly, if you had a
5-node cluster and you took down all 5-nodes and only restarted 3, the
cluster would start down-replicating ranges. The new behavior is to
consider a node part of the cluster until it is decommissioned. This
better matches user expectations.

Fixes #34122

Release note (bug fix): Avoid down-replicating widely replicated ranges
when nodes in the cluster are temporarily down.

petermattis committed:
roachtest: add replicate/wide roachtest

Add the `replicate/wide` roachtest which starts up a 9-node cluster,
sets the replication factor for all zones to 9, waits for full
replication, and then restarts the cluster, bringing up only nodes
1-6. Previously, this would cause down-replication and that
down-replication could cause unavailable ranges.

Further, test decommissioning one of the nodes and verify that the
replication of the ranges falls to 7. Lastly, decrease the replication
factor to 5 and verify the replicas per range again falls.

See #34122

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34126)
<!-- Reviewable:end -->

petermattis reviewed (commented):
I think we'll want to backport this PR to 2.1.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-:-LWcAW8887FTbA5XTgeS:b-dy08j4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/cmd/roachtest/allocator.go, line 377 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWcASEFDQ_bLGlj3AUj:-LWcASEFDQ_bLGlj3AUk:b-pu46uk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef1592f6379a12f3a19c5b13b532cfcb822c344/pkg/cmd/roachtest/allocator.go#L377)):*
> ```Go
> 
> 	// Set the replication width to 5. The replicas should down-replicate.
> 	// TODO(peter): This doesn't work. It should. There seems to be a bug.
> ```

I need to track down what is going on here. Just haven't had time yet.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/cmd/roachtest/allocator.go, line 377 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWcASEFDQ_bLGlj3AUj:-LWcmr0w7eff2FRYqG_9:b-6yzlb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef1592f6379a12f3a19c5b13b532cfcb822c344/pkg/cmd/roachtest/allocator.go#L377)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I need to track down what is going on here. Just haven't had time yet.
</blockquote></details>

`Allocator.RemoveTarget` won't remove a node for which we don't have a descriptor in the store pool. So in the scenario here in which we've restarted only 6-nodes of a 9-node cluster, those 6-nodes do not have store descriptors in their store pools for the 3 down nodes. After the time-until-store-dead threshold passes, the store pools start considering replicas on the down nodes as dead, which triggers a slightly different code path that simply removes a dead replica instead of calling into `Allocator.RemoveTarget`. This is somewhat surprising. It seems desirable to be able to remove a replica on a down-node before the time-until-store-dead threshold.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/cmd/roachtest/allocator.go, line 377 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWcASEFDQ_bLGlj3AUj:-LWgDgW55rcRDAlYxpL9:ba5n8lj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef1592f6379a12f3a19c5b13b532cfcb822c344/pkg/cmd/roachtest/allocator.go#L377)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`Allocator.RemoveTarget` won't remove a node for which we don't have a descriptor in the store pool. So in the scenario here in which we've restarted only 6-nodes of a 9-node cluster, those 6-nodes do not have store descriptors in their store pools for the 3 down nodes. After the time-until-store-dead threshold passes, the store pools start considering replicas on the down nodes as dead, which triggers a slightly different code path that simply removes a dead replica instead of calling into `Allocator.RemoveTarget`. This is somewhat surprising. It seems desirable to be able to remove a replica on a down-node before the time-until-store-dead threshold.
</blockquote></details>

I don't think there is anything to be done about this problem in the short term. Instead, I've adjusted `server.time_until_store_dead` to `90s` for this test. 


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
LGTM with some comments. Also, allocator changes always make me nervous, even if I know that they're fixing pretty bad bugs. That's a bad sign.

> Reviewed 1 of 3 files at r1, 14 of 14 files at r5, 2 of 2 files at r6, 10 of 10 files at r7, 2 of 2 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-:-LWpK_C_3HKckDkv4y5A:b-a9ywaq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @petermattis)

___
*[pkg/cmd/roachtest/allocator.go, line 377 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWcASEFDQ_bLGlj3AUj:-LWpBqu6C6oSyiMxN00j:b34keo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef1592f6379a12f3a19c5b13b532cfcb822c344/pkg/cmd/roachtest/allocator.go#L377)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I don't think there is anything to be done about this problem in the short term. Instead, I've adjusted `server.time_until_store_dead` to `90s` for this test. 
</blockquote></details>

so the allocator knows that it wants only 5 replicas and it has five healthy replicas and X more in unknown state, which it the refuses to remove until it knows the state? Hmm, that does seem suboptimal. Can you file an issue?
___
*[pkg/cmd/roachtest/allocator.go, line 353 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpK9Qu1UL0s-lOD0-n:-LWpK9Qu1UL0s-lOD0-o:bsvf3xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca9b29fae7934de975f8228325c61054031b9a1/pkg/cmd/roachtest/allocator.go#L353)):*
> ```Go
> 			}
> 			t.l.Printf("%d unavailable, %d under-replicated ranges\n", unavailable, underReplicated)
> 			if unavailable != 0 {
> ```

This has a theoretical bug in which it terminates too early because the metrics haven't been initialized yet. I don't think we write store statuses synchronously. You could assert on the count of rows.
___
*[pkg/cmd/roachtest/allocator.go, line 369 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpKPYiCbiRKEWNWDD6:-LWpKPYj5F_fE5e31aUZ:bvfabp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca9b29fae7934de975f8228325c61054031b9a1/pkg/cmd/roachtest/allocator.go#L369)):*
> ```Go
> 	decom := func(id int) {
> 		c.Run(ctx, c.Node(1),
> 			fmt.Sprintf("./cockroach node decommission --insecure --wait=none %d", id))
> ```

Now that I see this, it's a bit of a shame that we don't have a SQL syntax for decommissioning.
___
*[pkg/storage/allocator.go, line 1236 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWp43fO-UPbgKtlG2Gu:-LWp43fP9qu8XrdkHjvp:bsz4zo9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6bd16de492b4cff94e00b9985a1fd51a9efca9/pkg/storage/allocator.go#L1236)):*
> ```Go
> 	if len(upToDateReplicas) > quorum {
> 		// The number of up-to-date replicas is larger than quorum. Any replica can
> 		// be removed, though we want to filter out brandNewReplicaID
> ```

nit: missing dot.
___
*[pkg/storage/allocator_test.go, line 5243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWp4wRD7byRWoqAzhlQ:-LWp4wRECykkK8BZtWFe:bdzsyjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6bd16de492b4cff94e00b9985a1fd51a9efca9/pkg/storage/allocator_test.go#L5243)):*
> ```Go
> 		{3, []uint64{1, 2, 3, 4, 5}, 0, []uint64{1, 2}},
> 		{1, []uint64{1, 0}, 2, nil},
> 		{1, []uint64{1, 0}, 1, []uint64{0}},
> ```

Hmm, this is a bit dangerous, or am I reading this incorrectly? We have two replicas and are considering one removable? Even if we knew the other one was caught up that would lose quorum. Perhaps this is still the right behavior for this method.
PS reminds me of https://github.com/cockroachdb/cockroach/pull/32527.
___
*[pkg/storage/allocator_test.go, line 5246 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWp5JA-92xSENlzLgYg:-LWp5JA0C7a2DVpOmDp5:b-9dwn43) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6bd16de492b4cff94e00b9985a1fd51a9efca9/pkg/storage/allocator_test.go#L5246)):*
> ```Go
> 		{3, []uint64{3, 2, 1}, 3, nil},
> 		{3, []uint64{3, 2, 0}, 3, nil},
> 		{3, []uint64{4, 3, 2, 1}, 4, []uint64{2}},
> ```

I hope I don't understand well how this method is used. In this situation, I think it would be wise to wait until the last replica shows a better progress (or is known to have had its chance) to downreplicate. If the last replica is down, removing another one makes us unavailable.
___
*[pkg/storage/replica_init_test.go, line 46 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpA7ETCv0evspaUbRj:-LWpA7ETCv0evspaUbRk:b-8bnkyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e461e2872627ac3ba710a5d6ff41b5c037733e9c/pkg/storage/replica_init_test.go#L46)):*
> ```Go
> 		expectedLastReplicaAdded roachpb.ReplicaID
> 	}{
> 		// Adding a replica. Replica IDs always increase.
> ```

It looks like the code is immune to replica descs that don't have increasing replica IDs, which seems good. Do you want to test that too?
___
*[pkg/storage/store_pool.go, line 455 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpAqhe0F1p46-n4pEa:-LWpAqhe0F1p46-n4pEb:b-ar7i6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/50af1cbf73bfd63890ce185886e374ff36a274ab/pkg/storage/store_pool.go#L455)):*
> ```Go
> 
> // AvailableNodeCount returns the number of nodes that are possible allocation
> // targets. This includes dead nodes, but not decommissioning or decommissioned
> ```

Should this be called `ClusterNodeCount`? `Available` is ambiguous.
This method has a number of callers. Do they all want these semantics or do some of them not want to see dead nodes?


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @petermattis, and @tbg)

___
*[pkg/cmd/roachtest/allocator.go, line 377 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWcASEFDQ_bLGlj3AUj:-LWq2Esa77yj0KZCm7lF:b-geyexy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef1592f6379a12f3a19c5b13b532cfcb822c344/pkg/cmd/roachtest/allocator.go#L377)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

so the allocator knows that it wants only 5 replicas and it has five healthy replicas and X more in unknown state, which it the refuses to remove until it knows the state? Hmm, that does seem suboptimal. Can you file an issue?
</blockquote></details>

Yes, that's a correct summary. I'll file an issue.
___
*[pkg/cmd/roachtest/allocator.go, line 353 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpK9Qu1UL0s-lOD0-n:-LWq2QBx3OR3oJsVhqbX:b-q6izsw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca9b29fae7934de975f8228325c61054031b9a1/pkg/cmd/roachtest/allocator.go#L353)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This has a theoretical bug in which it terminates too early because the metrics haven't been initialized yet. I don't think we write store statuses synchronously. You could assert on the count of rows.
</blockquote></details>

 I don't think there is any problem if this code executes before the store statuses have been written since we're waiting for the number of under-replicated ranges to reach 20. 
___
*[pkg/cmd/roachtest/allocator.go, line 369 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpKPYiCbiRKEWNWDD6:-LWq2iGEEMw7qvSzuyvM:bc2czcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ca9b29fae7934de975f8228325c61054031b9a1/pkg/cmd/roachtest/allocator.go#L369)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Now that I see this, it's a bit of a shame that we don't have a SQL syntax for decommissioning.
</blockquote></details>

Ack. SQL-syntax-everywhere!
___
*[pkg/storage/allocator_test.go, line 5243 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWp4wRD7byRWoqAzhlQ:-LWq34Hi47b4Ef1RiAau:b-shlz4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6bd16de492b4cff94e00b9985a1fd51a9efca9/pkg/storage/allocator_test.go#L5243)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm, this is a bit dangerous, or am I reading this incorrectly? We have two replicas and are considering one removable? Even if we knew the other one was caught up that would lose quorum. Perhaps this is still the right behavior for this method.
PS reminds me of https://github.com/cockroachdb/cockroach/pull/32527.
</blockquote></details>

These test cases are super hard to read. I'd like to start introducing data-driven tests to the storage package. I think a lot of the allocator tests would be more readable, and I think we'd also uncover places where we're verifying suspicious behavior.

This scenario is that we have two replicas. The first is both caught up and considered a brand new replica (making it unremovable). The second is therefore the only removable candidate. Note that we're still going through the Raft membership change, so both replicas will need to be caught up in order to have quorum for this change. I looked into changing this when I first worked on this PR, but ran into a stumbling block with regards to moving single node replicas around in a cluster which is used by some tests. I'll add a TODO that this should be revisited.
___
*[pkg/storage/allocator_test.go, line 5246 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWp5JA-92xSENlzLgYg:-LWuIG-4DaCQUzypTf9v:babih4n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0d6bd16de492b4cff94e00b9985a1fd51a9efca9/pkg/storage/allocator_test.go#L5246)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I hope I don't understand well how this method is used. In this situation, I think it would be wise to wait until the last replica shows a better progress (or is known to have had its chance) to downreplicate. If the last replica is down, removing another one makes us unavailable.
</blockquote></details>

The calculation `filterUnremovableCandidates` is doing is funky. It is only allowing the removable of "behind" replicas if the number of up-to-date replicas equals quorum after down-replication. So in this instance, quorum after replication is 2, and 2 replicas are up-to-date, so it is allowing removal one of the other two. One of those replicas is "brand new" preventing removal. 

The funky bit, which doesn't seem right as I'm looking at it now, is to be using the quorum after down-replication. The down-replication isn't going to be able to succeed unless we have a quorum before down-replication.

Let me investigate this further.
___
*[pkg/storage/store_pool.go, line 455 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpAqhe0F1p46-n4pEa:-LWsRxxlBUvQug9zTd6C:bni9wbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/50af1cbf73bfd63890ce185886e374ff36a274ab/pkg/storage/store_pool.go#L455)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Should this be called `ClusterNodeCount`? `Available` is ambiguous.
This method has a number of callers. Do they all want these semantics or do some of them not want to see dead nodes?
</blockquote></details>

As far as I can tell, all of the callers of `AvailableNodeCount` feed the result into `GetNeededReplicas`. There is one exception: `TestCluster.WaitForFullReplication`. But there too, the cluster count is the desired count. 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @petermattis, and @tbg)

___
*[pkg/storage/store_pool.go, line 455 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpAqhe0F1p46-n4pEa:-LWufsyg5_sD1YHzmZ7m:b-io3w6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/50af1cbf73bfd63890ce185886e374ff36a274ab/pkg/storage/store_pool.go#L455)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

As far as I can tell, all of the callers of `AvailableNodeCount` feed the result into `GetNeededReplicas`. There is one exception: `TestCluster.WaitForFullReplication`. But there too, the cluster count is the desired count. 
</blockquote></details>

Ok. I still feel that the name could be clearer.


<!-- Sent from Reviewable.io -->

petermattis committed:
storage: rename AvailableNodeCount to ClusterNodeCount

Release note: None

petermattis committed:
storage: disallow down-replication on below quorum ranges

Change `filterUnremovableCandidates` to return zero candidates if the
range is currently below quorum. This likely has little effect in
practice as a below quorum range would be unable to process a replica
removal.

Release note: None

petermattis reviewed (commented):
>  Also, allocator changes always make me nervous, even if I know that they're fixing pretty bad bugs. That's a bad sign.

I think the allocator code could (should) be refactored into a separate package. And as I noted elsewhere, I think this code would benefit from data-driven tests to give us more confidence in the test cases (they could become significantly more readable).

I think I've addressed all of your comments. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-:-LWv46xl3JqVCu6y31T4:b5p8ehs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/storage/replica_init_test.go, line 46 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpA7ETCv0evspaUbRj:-LWv3rpl-DKwns9khysO:bf1qs8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e461e2872627ac3ba710a5d6ff41b5c037733e9c/pkg/storage/replica_init_test.go#L46)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

It looks like the code is immune to replica descs that don't have increasing replica IDs, which seems good. Do you want to test that too?
</blockquote></details>

Done. The code leaves `lastReplicaAdded` at zero, which seems fine given this shouldn't happen in practice.
___
*[pkg/storage/store_pool.go, line 455 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-LWpAqhe0F1p46-n4pEa:-LWusS3mEK3f4q7kz5rd:b-jkisg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/50af1cbf73bfd63890ce185886e374ff36a274ab/pkg/storage/store_pool.go#L455)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Ok. I still feel that the name could be clearer.
</blockquote></details>

Renamed to `ClusterNodeCount` and all of the variables from `availableNodes` to `clusterNodes`.


<!-- Sent from Reviewable.io -->

tbg commented:
Looks like you have a merge conflict.

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> and thanks for digging into these.

> Reviewed 14 of 14 files at r9, 2 of 2 files at r10, 10 of 10 files at r11, 8 of 8 files at r12, 2 of 2 files at r13, 2 of 2 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34126#-:-LWvAqAr41U8CF0GLxN1:brlgzw2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)



<!-- Sent from Reviewable.io -->

petermattis commented:
Merge conflict fixed. TFTR!

petermattis commented:
bors r=tbg
